---
title: "iSEE_and_iSEEfier_Workshop"
author: "Najla Abassi[^najla.abassi@uni-mainz.de]"  
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{iSEE_and_iSEEfier_Workshop}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
collapse = TRUE,
comment = "#>"
)
```

# Streamlining Omics Data Visualization with `iSEE` and `iSEEfier`

Workshop adapted from the collection of material developed by the iSEE core dev team

Authors: Najla Abassi[^workshop_iSEEfier-2]. <br/> Last modified: "`r Sys.Date()`"

[^workshop_iSEEfier-2]: Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), University Medical Center, Mainz, Germany

## Introduction to `iSEE`

### Input data

`r BiocStyle::Biocpkg("iSEE")` was designed around the `r BiocStyle::Biocpkg("SummarizedExperiment")` class, a container widely used throughout the *Bioconductor* project. Briefly, the `r BiocStyle::Biocpkg("SummarizedExperiment")` class provides a container keeping matrices of assay data, sample metadata, and feature metadata. ![SummarizedExperiment (source: the package vignette; \<https://bioconductor.org/packages/SummarizedExperiment/\\\\\\\>).](img/SummarizedExperiment.png)

Now we will start by loading a demonstration dataset of single cell RNAseq dataset from Tasic et al. (2016) "Adult mouse cortical cell taxonomy revealed by single cell transcriptomics" [doi:10.1038/nn.4216](https://doi.org/10.1038/nn.4216). The data is made available via the `r BiocStyle::Biocpkg("scRNAseq")` Bioconductor package, and is stored in an object of class `r BiocStyle::Biocpkg("SingleCellExperiment")`, which is a class that inherits from the `SummarizedExperiment` class. It includes additional slots and methods like metadata related to individual cells and their associated dimensionality reductions. We processed the data beforehand to save some time.

```{r process_data, eval=FALSE, echo=FALSE}
# load data
sce <- scRNAseq::ReprocessedAllenData(assays="tophat_counts")
# rename tophat_counts to counts
assayNames(sce)[assayNames(sce) == "tophat_counts"] <- "counts"
# log transform the counts
sce <- scuttle::logNormCounts(sce)
# add dimensionality reduction coordinates
sce <- scater::runPCA(sce)
sce <- scater::runTSNE(sce)
sce <- scater::runUMAP(sce)
saveRDS(sce,"sce_allen.RDS")
```

```{r loadlib, message=FALSE,warning=FALSE}
library("iSEEfierWorkshopSCBS2024")
```

```{r loaddata}
sce <- readRDS("data/sce_allen.RDS")
sce
```

Looking at the object displayed above, we can tell that it contains:

-   Two `assays`. The first one called `counts`, and corresponds to a matrix measuring 20,816 genes in 379 cells. And The second assay is `logcounts`, which is the log transformed counts.
-   Twenty three columns of cell metadata (`colData`).
-   No gene metadata (`rowData`).
-   Three reduced dimensions : `PCA`, `TSNE`, and `UMAP`.

Also of note:

-   Gene symbols are in use for the row names.
-   Arbitrary cell names are in use for column names.
-   Some object-level metadata is present.
-   An alternative experiment called `ERCC` is present (we will not use it).

### The default `iSEE` app

Now we can launch an `r BiocStyle::Biocpkg("iSEE")` instance and start exploring this dataset using the `iSEE()` function without any further argument. This will produce an app using the default configuration; that is, the app instance will include one panel of each built-in class for which the relevant information is available in the `SingleCellExperiment` object.

```{r createApp}
app <- iSEE::iSEE(sce)
```

```{r launchApp, message=FALSE, eval=FALSE}
runApp(app,launch.browser = TRUE)
```

```{r defaultApp, echo=FALSE, out.width='100%'}
webshot2::appshot(app,"img/vanilla-isee.png", delay = 10)
```

# The built-in panel types

## `ReducedDimensionPlot`

The reduced dimension plot can display any reduced dimension representation that is present in the `reducedDim` slot of the `SingleCellExperiment` object.

Note that this slot is not defined for the base `SummarizedExperiment` class, in which case the user interface does not allow the inclusion of panels of this type.

```{r reddim, message=FALSE, echo = FALSE}
app <- iSEE::iSEE(sce, initial = list(ReducedDimensionPlot()))
```

```{r reddimplot, echo=FALSE, out.width='100%'}
webshot2::appshot(app,"img/isee-reduceddimensionplot.png", delay = 10)
```

## `ColumnDataPlot`

The column data plot can display one or two of the provided column annotations (from the `colData` slot). Depending on the class of the selected annotations, the panel shows either a scatter plot, a violin plot, or a Hinton diagram [@Hinton1991-hintondiagram; @Bremner1994-hintonplots].

```{r coldata, message=FALSE, echo = FALSE}
app <- iSEE::iSEE(sce, initial = list(ColumnDataPlot()))
```

```{r coldataplot, echo=FALSE, out.width='100%'}
webshot2::appshot(app, "img/isee-columndataplot.png", delay = 10)
```

## `RowDataPlot`

Analogous to the [column data plot](#coldataplot) above, the row data plot displays one or two of the provided row annotations (from the `rowData` slot). Depending on the class of the selected annotations, the panel displays either a scatter plot, a violin plot, or a Hinton plot.

```{r rowdata, message=FALSE, echo = FALSE}
app <- iSEE::iSEE(sce, initial = list(RowDataPlot(PanelWidth = 12L)))
```

```{r rowdataplot, echo=FALSE, out.width='100%'}
webshot2::appshot(app, "img/isee-rowdataplot.png", delay = 10)
```

## `ComplexHeatmapPlot`

The complex heatmap panel displays, for any assay, the observed values for a subset of the features across the samples.

```{r heatmap, message=FALSE, echo = FALSE}
app <- iSEE::iSEE(sce, initial = list(ComplexHeatmapPlot(PanelWidth = 12L)))
```

```{r cmplxHeatmap, echo=FALSE, out.width='100%'}
webshot2::appshot(app, "img/isee-complexheatmapplot.png", delay = 10)
```

## `FeatureAssayPlot`

The feature assay plot displays the observed values for one feature across the samples. It is also possible to plot the observed values for two features, in a scatter plot.

```{r featureassay, message=FALSE, echo = FALSE}
app <- iSEE::iSEE(sce, initial = list(FeatureAssayPlot(PanelWidth = 12L)))
```

```{r featureassayplot, echo=FALSE, out.width='100%'}
webshot2::appshot(app, "img/isee-featureassayplot.png", delay = 10)
```

## `SampleAssayPlot`

Analogous to the [`r .fullName(FeatureAssayPlot())`](#featassayplot) above, the `r .fullName(SampleAssayPlot())` shows the observed values for all features, for one of the samples. It is also possible to plot the observed values for two samples, in a scatter plot.

```{r sampleassay, message=FALSE, echo = FALSE}
app <- iSEE::iSEE(sce, initial = list(SampleAssayPlot(PanelWidth = 12L)))
```

```{r sampleassayplot, echo=FALSE, out.width='100%'}
webshot2::appshot(app, "img/isee-sampleassayplot.png", delay = 10)
```

## `RowDataTable` {#rowdatatable}

The row data table displays all information provided in the `rowData` slot of the `SummarizedExperiment` object, leveraging the interactivity provided by the `r BiocStyle::CRANpkg("DT")` package.

```{r rowdatatab, message=FALSE, echo = FALSE}
app <- iSEE::iSEE(sce, initial = list(RowDataTable(PanelWidth = 12L)))
```

```{r rowdatatab1, echo=FALSE, out.width='100%'}
webshot2::appshot(app, "img/isee-rowdatatable.png", delay = 10)
```

## `ColumnDataTable`

Analogous to the [`r .fullName(RowDataTable())`](#rowdatatable) above, the column data table displays all information provided in the `colData` slot of the `SummarizedExperiment` object.

```{r coldatatab, message=FALSE, echo = FALSE}
app <- iSEE::iSEE(sce, initial = list(ColumnDataTable(PanelWidth = 12L)))
```

```{rcoldatatab1, echo=FALSE, out.width='100%'}
webshot2::appshot(app, "img/isee-columndatatable.png", delay = 10)
```

# Collapsible boxes with display controls

## Data parameters

Each plot panel type has a `Data parameters` collapsible box. This box has different content for each panel type, but in all cases it lets the user control the data that is displayed in the plot.

```{r datapram, message=FALSE, echo = FALSE}
app <- iSEE::iSEE(sce, initial = list(ColumnDataPlot(PanelWidth = 12L)))
```

```{r dataparam1, echo=FALSE, out.width='100%'}
webshot2::appshot(app, "img/isee-columndataplot-dataparameters.png", delay = 10)
```

## Visual parameters

In contrast to the `Data parameters` collapsible box that lets users control *what* is displayed in the plot, the `Visual parameters` box lets users control *how* the information is displayed.

This collapsible box contains the controls to change the size, shape, opacity, and color of the points, to facet the plot by any available categorical annotation, to subsample points for increased speed of plot rendering, and to control how legends are displayed.

```{r vizparam, message=FALSE, echo=FALSE}
app <- iSEE::iSEE(sce, initial = list(ColumnDataPlot(PanelWidth = 12L)))
```

```{r vizparam1, echo=FALSE, out.width='100%'}
webshot2::appshot(app, "img/isee-columndataplot-visualparameters.png", delay = 10)
```

## Selection parameters

The `Selection parameters` collapsible box provides controls to transfer selections of points (features or samples) between panels.

We demonstrate examples of point transmission in the separate vignette of [workshop recipes](recipes.html)

```{r selecparam, message=FALSE, echo = FALSE}
app <- iSEE::iSEE(sce, initial = list(ReducedDimensionPlot(PanelWidth = 12L)))
```

```{r selecparam1, echo=FALSE, out.width='100%'}
webshot2::appshot(app, "img/isee-reduceddimensionplot-selectionparameters.png", delay = 10)
```

# Additional controls

At the top-right corner of the iSEE application, users can find additional controls for reproducibility, configuration, and help.

```{r addcontrol, echo = FALSE, eval = TRUE}
panel_organization_fig <- "img/PanelOrganization.png"
export_fig <- "img/Export.png"
documentation_fig <- "img/Documentation.png"
additional_info_fig <- "img/AdditionalInfo.png"
```

## Introduction to `iSEEfier`

Now we have seen how powerful of a tool iSEE can be. However

# iSEEfier

Let's say we are interested in visualizing the expression of a list of specific marker genes in one view, or maybe we created different initial states separately, but would like to visualize them in the same instance. As we previously learned, we can do a lot of these tasks by running the command:

```{r, eval=FALSE}
# don't run
iSEE(sce)
```

Then, add/remove the panels to our need. This can involve multiple steps (selecting the gene of interest, color by a specific `colData`...), or probably even write multiple lines of code.
For that we can use the `r Biocpkg("iSEEfier")` package, which streamlines the process of starting (or if you will, firing up) an `iSEE` instance with a small chunk of code, avoiding the tedious way of setting up every `iSEE` panel individually.

In this section, we will illustrate a simple example of how to use `r Biocpkg("iSEEfier")`. We will use the same pbmc3k data we worked with during this workshop.

We start by loading the data:

```{r, eval=FALSE}
library("iSEEfier")

# import data
sce <- readRDS(
  file = system.file("datasets", "sce_pbmc3k.RDS", package = "iUSEiSEE")
)

sce
```

For example, we can be interested in visualizing the expression of *GZMB*, *TGFB*, and *CD28* genes all at once. We start by providing a couple of parameters:

```{r, eval=FALSE}
# define the list of genes
feature_list_1 <- c("GZMB", "TGFB1", "CD28")
# define the cluster/cell type 
cluster_1 <- "labels_main"
```

Now we can pass these parameters into `iSEEinit()` to create a customized initial configuration:

```{r, eval=FALSE}
# create an initial state with iSEEinit
initial_1 <- iSEEinit(sce,
                      features = feature_list_1,
                      clusters = cluster_1,
                      add_markdown_panel = TRUE)
```

The rest can be as easy as passing this initial to the `iSEE()` call:

```{r, eval=FALSE}
app <- iSEE(sce, initial = initial_1)
```

This is how it would look like:

```{r, echo=FALSE, out.width='100%', eval=FALSE}
SCREENSHOT("images/isee-on-fier.png", delay = 30)
```


While we are visualizing the expression of these genes, we might want to take some notes (gene X is more expressed in a certain cell type/cluster than some others, maybe we are trying to annotate the cells ourselves if the annotation wasn't available...).  
For this, we used the argument `add_markdown_panel = TRUE`. It will display a `MarkdownBoard` panel where we can note our observations without leaving the app.

We can also use the argument `add_dynamicTable_panel=TRUE` to add another custom panel to display the maker genes of certain cluster/cell type.

```{r, eval=FALSE}
feature_list_2 <- c("CD74", "CD79B")
initial_2 <- iSEEinit(sce,
                      features = feature_list_2,
                      clusters = cluster_1,
                      add_markdown_panel = TRUE,
                      add_dynamicTable_panel = TRUE)
```

```{r, eval=FALSE}
app <- iSEE(sce, initial = initial_2)
```

We can check the initial's content, or how the included panels are linked between each other without running the app with `view_initial_tiles()` and `view_initial_network()`:

```{r, eval=FALSE}
# display a graphical representation of the initial configuration, where the panels are identified by their corresponding colors
view_initial_tiles(initial = initial_1)
# display a network visualization for the panels
view_initial_network(initial_1, plot_format = "igraph")
```

Another alternative for network visualization would use the interactive widget provided by `visNetwork`:

```{r, eval=FALSE}
view_initial_network(initial_1, plot_format = "visNetwork")
```

It is also possible to combine multiple initials into one:

```{r, eval=FALSE}
merged_config <- glue_initials(initial_1, initial_2)
# check out the content of merged_config
view_initial_tiles(initial = merged_config)
```

`?iSEEfier` is always your friend whenever you need further documentation on the package/a certain function and how to use it.

## Session info {.unnumbered}

```{r}
sessioninfo::session_info()
```
